core/state/database.go

```diff
@@ -136,21 +136,151 @@ func NewDatabase(db ethdb.Database) Database {
 // large memory cache.
 func NewDatabaseWithConfig(db ethdb.Database, config *trie.Config) Database {
 	csc, _ := lru.New(codeSizeCacheSize)
-	if config != nil && config.UseVerkle {
-		return &VerkleDB{
+	return &ForkingDB{
+		cachingDB: &cachingDB{
+			db:            trie.NewDatabaseWithConfig(db, config),
+			disk:          db,
+			codeSizeCache: csc,
+			codeCache:     fastcache.New(codeCacheSize),
+		},
+		VerkleDB: &VerkleDB{
 			db:            trie.NewDatabaseWithConfig(db, config),
 			diskdb:        db,
 			codeSizeCache: csc,
 			codeCache:     fastcache.New(codeCacheSize),
 			addrToPoint:   utils.NewPointCache(),
+		},
+		started: (config != nil && config.UseVerkle),
+		ended:   (config != nil && config.UseVerkle),
+	}
+}
+
+// ForkingDB is an adapter object to support forks between
+// cachingDB and VerkleDB.
+type ForkingDB struct {
+	*cachingDB
+	*VerkleDB
+
+	started, ended  bool
+	translatedRoots map[common.Hash]common.Hash // hash of the translated root, for opening
+
+	// TODO ensure that this info is in the DB
+	LastAccHash  common.Hash
+	LastSlotHash common.Hash
+}
+
+// ContractCode implements Database
+func (fdb *ForkingDB) ContractCode(addrHash common.Hash, codeHash common.Hash) ([]byte, error) {
+	if fdb.started {
+		return fdb.VerkleDB.ContractCode(addrHash, codeHash)
+	}
+
+	return fdb.cachingDB.ContractCode(addrHash, codeHash)
+}
+
+// ContractCodeSize implements Database
+func (fdb *ForkingDB) ContractCodeSize(addrHash common.Hash, codeHash common.Hash) (int, error) {
+	if fdb.started {
+		return fdb.VerkleDB.ContractCodeSize(addrHash, codeHash)
+	}
+
+	return fdb.cachingDB.ContractCodeSize(addrHash, codeHash)
+}
+
+// CopyTrie implements Database
+func (fdb *ForkingDB) CopyTrie(t Trie) Trie {
+	mpt := fdb.cachingDB.CopyTrie(t)
+	overlay := fdb.VerkleDB.CopyTrie(t)
+
+	if fdb.started {
+		return trie.NewTransitionTree(mpt.(*trie.SecureTrie), overlay.(*trie.VerkleTrie))
+	}
+
+	return mpt
+}
+
+// OpenStorageTrie implements Database
```

Comment:
So this implementation is somewhat tacky: there are several trees in the MPT world, but only one in the verkle world. I usually worked around the problem by making sure this would never get called, and it's probably the right approach still.

But I'm also thinking that this function should disappear from the interface in upstream altogether: one can see all these trees as only one tree, after all. I need to think it over, but this hack should work for now.